scoping bug - having variable exist in outer scope change value used inside for loop

```go
// scoping bug
ext := "X" // commenting this out hides the bug
list := ["A","B","C"]
l2 := [1,2]
func testScoping() {
	res := []
	for ext := list {
		for idx2 := l2 {
			res = res + [[idx2, ext]]
		}
	}
	return res
}
res := testScoping()
```
expect
```go
[[1,"A"],[2,"A"],[1,"B"],[2,"B"],[1,"C"],[2,"C"]]
```
actual
```go
[[1,"C"],[2,"C"],[1,"C"],[2,"C"],[1,"C"],[2,"C"]]
```
(!)

reduced to
```
// scoping bug
ext := "X" // commenting this out hides the bug
list := ["A","B"]
func testScoping(i) {
	res := []
	for ext := list {
		res = res + [ext]
	}
	return res
}

res1 := testScoping("a") // should be ["A","B"] gets ["B", "B"]
log("res1:", res1)
del(ext)
res2 := testScoping("b")
log("res2:", res2)
Assert("for scoping bug", res1 == ["A", "B"] && res2 == ["A", "B"])
log("not reached")
```
even more fun that addition of the del() cause segv! 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x104828964]

goroutine 1 [running]:
grol.io/grol/object.Cmp({0x104934698?, 0x140001d05d0?}, {0x104934628, 0x140001d3380})
	grol.io/grol/object/object.go:164 +0x54
grol.io/grol/object.Cmp({0x104934740?, 0x140001c3b00?}, {0x104934740, 0x140001d82d0})
	grol.io/grol/object/object.go:222 +0xa8c
grol.io/grol/object.Equals({0x104934740, 0x140001c3b00}, {0x104934740, 0x140001d82d0})
	grol.io/grol/object/object.go:133 +0x8c
grol.io/grol/eval.(*State).evalInfixExpression(0x1400010a480, 0x27, {0x104934740, 0x140001c3b00}, {0x104934740, 0x140001d82d0})
	grol.io/grol/eval/eval.go:1156 +0x668
grol.io/grol/eval.(*State).evalInternal(0x1400010a480, {0x1049107e0?, 0x140001d61b0})
	grol.io/grol/eval/eval.go:249 +0x5bc
grol.io/grol/eval.(*State).Eval(0x1400010a480, {0x1049107e0?, 0x140001d61b0?})
	grol.io/grol/eval/eval_api.go:151 +0x44
grol.io/grol/eval.(*State).evalInternal(0x1400010a480, {0x1049107e0?, 0x140001d6180})
	grol.io/grol/eval/eval.go:230 +0x2fc
```

reduced to
```go
// scoping bug

list := ["A","B"]
ext := "X" // commenting this out hides the bug

// -- segv issue: --
func testScoping2() {
	res := []
	for ext := list {
		res = res + [ext]
	}
	return res
}
res1 := testScoping2()
log("res1:", res1)
del(ext)
res2 := testScoping2()
log("res2:", res2)
Assert("for scoping bug", res1 == ["A", "B"]) // segv
log("done with scoping bug")

// -- regular scoping bug --

func testScoping(i) {
	res := []
	for ext := list {
		res = res + [ext]
	}
	return res
}

res3 := testScoping("a") // should be ["A","B"] gets ["B", "B"]
log("res3:", res3)
ext := "X" // commenting this out hides the bug
res4 := testScoping("b")
log("res4:", res4)
Assert("for scoping bug", res3 == ["A", "B"] && res4 == ["A", "B"])
```

--- a/eval/eval.go
+++ b/eval/eval.go
@@ -1008,7 +1008,10 @@ func (s *State) evalForList(fe *ast.ForExpression, list object.Object, name stri
 		if v == nil {
 			return s.NewError("for list element is nil")
 		}
-		s.env.Set(name, v)
+		oerr := s.env.CreateOrSet(name, v, true) // Create new local scope for loop variable
+		if oerr.Type() == object.ERROR {
+			return oerr
+		}
 		// Copy pasta from evalForInteger. hard to share control flow.
 		nextEval := s.evalInternal(fe.Body)
 		switch nextEval.Type() {

--- a/object/object.go
+++ b/object/object.go
@@ -764,6 +764,10 @@ func NewArray(elements []Object) Object {
 	if len(elements) == 0 {
 		return EmptyArray
 	}
+	// Dereference values to current values vs keeping references to captured variables.
+	for i := range elements {
+		elements[i] = Value(elements[i])
+	}
 	if len(elements) <= MaxSmallArray {
 		sa := SmallArray{len: len(elements)}
 		copy(sa.smallArr[:], elements)
@@ -1198,10 +1202,14 @@ type Reference struct {
 }
 
 func (r Reference) ObjValue() Object {
-	if log.LogDebug() {
-		log.Debugf("Reference Value() %s -> %s", r.Name, r.RefEnv.store[r.Name].Inspect())
+	v, ok := r.RefEnv.store[r.Name]
+	if !ok {
+		// Reference points to a deleted variable
+		return Error{Value: "reference to deleted variable " + r.Name}
 	}
-	v := r.RefEnv.store[r.Name]
+	if log.LogDebug() {
+		log.Debugf("Reference Value() %s -> %s", r.Name, v.Inspect())
+	}
 	if v == r {
 		panic("Self reference")
 	}

--- a/object/state.go
+++ b/object/state.go
@@ -265,7 +265,8 @@ func (e *Environment) Get(name string) (Object, bool) {
 }
 
 // Delete removes the first entry found under that name from the environment.
-// TODO: check if references need special handling.
+// TODO: check if references need special handling / if variable is in use/referenced somewhere.
+// (had a SEGV in https://github.com/grol-io/grol/issues/335 ).
 func (e *Environment) Delete(name string) Object {
 	if e.depth == 0 {
 		e.numSet++

--- a/tests/for.gr
+++ b/tests/for.gr
@@ -68,3 +68,18 @@ ext := "X" // commenting this out hides the bug
 res4 := testScoping("b")
 log("res4:", res4)
 Assert("for scoping bug", res3 == ["A", "B"] && res4 == ["A", "B"])
+
+// -- reference deletion with for loop test --
+list := ["A", "B"]
+ext := "X"
+func testForRefDelete() {
+    res := []
+    for x := list {
+        res = res + [[x, ext]]
+    }
+    return res
+}
+
+res := testForRefDelete()
+del(ext)
+Assert("for ref delete", res == [["A", "X"], ["B", "X"]])

--- a/tests/for.gr
+++ b/tests/for.gr
@@ -29,3 +29,42 @@ func testRef() {
 }
 
 NoErr("for global array inside a function", testRef(), "6")
+
+// scoping bug
+
+list := ["A","B"]
+ext := "X" // commenting this out hides the bug
+
+// -- segv issue: --
+func testScoping2() {
+	res := []
+	for ext := list {
+		res = res + [ext]
+	}
+	return res
+}
+res1 := testScoping2()
+log("res1:", res1)
+del(ext)
+res2 := testScoping2()
+log("res2:", res2)
+Assert("for scoping bug", res2 == ["A", "B"])
+Assert("for scoping bug", res1 == ["A", "B"]) // segv comment out to see next test error
+log("done with scoping bug")
+
+// -- regular scoping bug --
+
+func testScoping(i) {
+	res := []
+	for ext := list {
+		res = res + [ext]
+	}
+	return res
+}
+
+res3 := testScoping("a") // should be ["A","B"] gets ["B", "B"]
+log("res3:", res3)
+ext := "X" // commenting this out hides the bug
+res4 := testScoping("b")
+log("res4:", res4)
+Assert("for scoping bug", res3 == ["A", "B"] && res4 == ["A", "B"])

